nightingale_perception/src/nightingale_perception/pose_estimation.py

```diff
@@ -0,0 +1,392 @@
+        self.ee_bin_adjustment = 0.0  # meters
+
+    def depth_image_cb(self, data):
+        # store the latest depth image for processing
+        try:
+            self.last_depth_img = self.cv_bridge.imgmsg_to_cv2(data, "16UC1")
+        except CvBridgeError as e:
+            rospy.loginfo(e)
+
+    def rgb_image_cb(self, data):
+        # store the latest color image for processing
+        try:
+            self.last_rgb_img = self.cv_bridge.imgmsg_to_cv2(data, "bgr8")
+        except CvBridgeError as e:
+            rospy.loginfo(e)
+
+    def camera_info_cb(self, data):
+        # plumb model is also known as brown conrady
+        self.depth_optical_link_frame = data.header.frame_id
+        if data.distortion_model == "plumb_bob":
+            self.camera_intrinsics.model = rs.distortion.brown_conrady
+        else:
+            self.camera_intrinsics.model = rs.distortion.none
+        # intrinsics given as 1D array
+        self.camera_intrinsics.fx = data.K[0]
+        self.camera_intrinsics.fy = data.K[4]
+        self.camera_intrinsics.ppx = data.K[2]
+        self.camera_intrinsics.ppy = data.K[5]
+        self.camera_intrinsics.width = data.width
+        self.camera_intrinsics.height = data.height
+        rospy.loginfo(f"Got camera intrinsics {self.camera_intrinsics}. Unregistering")
+        # only need to get camera info once
+        self.camera_info_sub.unregister()
+
+    def execute_cb(self, goal):
+        rospy.loginfo(f"Recieved target {goal.target}. Beginning pose estimation")
+
+        with mp.solutions.pose.Pose(
+            static_image_mode=True,
+            model_complexity=2,
+            enable_segmentation=True,
+            min_detection_confidence=0.5,
+        ) as pose:
+            self.last_rgb_img.flags.writeable = False
+            image = cv2.cvtColor(self.last_rgb_img, cv2.COLOR_BGR2RGB)
+            results = pose.process(image)
+            landmarks_list = results.pose_landmarks.landmark
+
+            if goal.target == "body":
+                bin_goal = self.pose_estimate_body(landmarks_list)
+                # should not be entered as of yet
+                if bin_goal is None:
+                    rospy.loginfo("Pose estimate failed")
+                    self.result.pose_est_fail = 1
+                    self.pose_estimation_server.set_aborted(self.result)
+                    return
+                self.result.bin_goal = bin_goal
+                self.pose_estimation_server.set_succeeded(self.result)
+            elif goal.target == "head":
+                rospy.loginfo(f"Head not implemented")
+                self.result.pose_est_fail = 2
+                self.pose_estimation_server.set_aborted(self.result)
+                return
+            else:
+                rospy.loginfo(f"Need valid target")
+                self.result.pose_est_fail = 3
+                self.pose_estimation_server.set_aborted(self.result)
+
+    def preempt_cb(self):
+        pass
```

Comment:
raise NotImplemented

Reply:
I realized this node returns so fast that it's pretty much a service so I don't think this will ever be called